### PR TITLE
LA-19 - Create Lambda service for xAPI statement ingestion

### PR DESCRIPTION
**Pull request by sandeepmjay**
Created an AWS Lambda micro-service to ingest xAPI statement into LRS. The service authenticates user, validates xAPI statement and writes to LRS, if the statement doesn't already exist and is in conformance to the xAPI standard.

### CODE REVIEW

File: xapi-ingest-microservice/api.js [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+ *
+ * @param  {Object}           ctx                       Context containing write credentials
+ * @param  {Object}           statement                 The learning activity statement to save
+ * @param  {Function}         callback                  Standard callback function
+ * @param  {Object}           callback.err              An error that occurred, if any
+ */
+var saveStatement = module.exports.saveStatement = function(ctx, statement, callback) {
+  if (!ctx || !ctx.auth) {
+    console.log({'err': err}, 'Prevented storing a learning activity without authentication');
+    return callback({'code': 500, 'msg': 'Prevented storing a learning activity without authentication'});
```
Comment (johncrossman) [inline xapi-ingest-microservice/api.js]: @sandeepmjay would it be useful to log `statement` id or similar?

File: xapi-ingest-microservice/api.js [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+  }
+
+  // Verify that no other learning activity statement with the same uuid already exists
+  if (statement.id) {
```
Comment (johncrossman) [inline xapi-ingest-microservice/api.js]: @sandeepmjay do nothing if `statement.id` is null?

File: xapi-ingest-microservice/api.js [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+      } else if (err && err.code !== 404) {
+        console.log({'id': statement.id}, 'Unable to verify if learning activity statement already exists');
+        return callback(err);
+      }
+
+      // Validate the statement
+      xapiValidator.validate(statement, function(err) {
+        if (err) {
+          console.log({'err': err}, 'Invalid learning activity statement');
+          return callback({'code': 500, 'msg': err});
```
Comment (johncrossman) [inline xapi-ingest-microservice/api.js]: @sandeepmjay I would expect `'msg': err.message` per convention. No biggie.

File: xapi-ingest-microservice/api.js [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+        }
+
+        // When no uuid has been included, generate one
+        statement.id = statement.id || uuid.v4();
```
Comment (johncrossman) [inline xapi-ingest-microservice/api.js]: @sandeepmjay given `if (statement.id)` at line 82, the line above will never use `uuid.v4()`

File: xapi-ingest-microservice/api.js [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+          } else if (refStatement && refStatement.object && refStatement.object.definition && refStatement.object.definition.type) {
+            activityType += '_' + refStatement.object.definition.type.split('/').pop();
+          }
+          activityType = activityType.toLowerCase();
+
+          // Get the user associated to this learning activity. If the user
+          // doesn't exist yet, it will be created
+          getOrCreateUser(ctx, statement, function(err, user) {
+            if (err) {
+              console.log({'err': err}, 'An error occured while getting the learning activity statement actor');
```
Comment (johncrossman) [inline xapi-ingest-microservice/api.js]: @sandeepmjay typo: should be "occurred"

File: xapi-ingest-microservice/api.js [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+              'write_credential_id': ctx.auth.id
+            };
+
+            DB.Statement.create(storedStatement).complete(function(err, statement) {
+              if (err) {
+                console.log({'err': err}, 'Failed to store a new learning activity statement');
+                return callback({'code': 500, 'msg': err.message});
+              }
+
+              console.log({'statement': statement}, 'Sucessfully stored learning activity statement');
```
Comment (johncrossman) [inline xapi-ingest-microservice/api.js]: @sandeepmjay typo: "Sucessfully"

File: xapi-ingest-microservice/api.js [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+ *
+ * @param  {Object}           statement                 The learning activity statement to check for referenced learning activities
+ * @param  {Function}         callback                  Standard callback function
+ * @param  {Object}           callback.err              An error that occurred, if any
+ * @param  {Statement}        callback.refStatement     The learning activity statement referenced in the provided learning activity statement
+ * @api private
+ */
+var getStatementRef = function(statement, callback) {
+  // The statement doesn't contain a statement reference
+  if (!statement.object || !(statement.object.objectType === 'StatementRef' && statement.object.id)) {
```
Comment (johncrossman) [inline xapi-ingest-microservice/api.js]: @sandeepmjay this conditional looks right but it's hard for my tired brain to parse.  Is the following equivalent and more readable?
```
if (!statement.object || statement.object.objectType !== 'StatementRef' || !statement.object.id) {
```

File: xapi-ingest-microservice/api.js [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+    return callback();
+  }
+
+  getStatement(statement.object.id, function(err, refStatement) {
+    if (err && err.code !== 404) {
+      console.log({'id': id}, 'Could not find a referenced learning activity statement');
```
Comment (johncrossman) [inline xapi-ingest-microservice/api.js]: @sandeepmjay do you mean `{'id': statement.object.id}`?  And why do we ignore the 404 case?

File: xapi-ingest-microservice/index.js [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+  require('./globals/sequelize.js');
+  console.log('Starting DB init function');
+  DB.init(function() {
+    AuthAPI.verifyWriteAuth(context, event, function(err) {
+      if (err) {
+        context.fail(err.code + ' : ' + err.msg);
+      }else{
+        console.log('Connection Successful !!');
+        console.log('Event :' + JSON.stringify(event));
+        console.log('Context' + JSON.stringify(context));
```
Comment (johncrossman) [inline xapi-ingest-microservice/index.js]: @sandeepmjay a few minor spacing issues above.

File: xapi-ingest-microservice/api.js (hunk 7 above)
Comment (sandeepmjay) [inline xapi-ingest-microservice/api.js]: The function checks if the a referenced learning activity is already available in the LRS. Hence, the combined condition to check if Object Type = StatementRef and there is an Id associated with it. The retrieval of the statement won't be possible without the Id. 

Also, the condition reappears in xapi-caliper utility and the main LRS application. Hence, keeping it consistent.

File: xapi-ingest-microservice/api.js (hunk 5 above)
Comment (sandeepmjay) [inline xapi-ingest-microservice/api.js]: Fixed.

File: xapi-ingest-microservice/api.js (hunk 6 above)
Comment (sandeepmjay) [inline xapi-ingest-microservice/api.js]: Fixed

File: xapi-ingest-microservice/index.js (hunk 9 above)
Comment (sandeepmjay) [inline xapi-ingest-microservice/index.js]: Fixed.

File: xapi-ingest-microservice/api.js (hunk 3 above)
Comment (sandeepmjay) [inline xapi-ingest-microservice/api.js]: Improving overall logging in next iteration.

File: xapi-ingest-microservice/api.js (hunk 1 above)
Comment (sandeepmjay) [inline xapi-ingest-microservice/api.js]: Will be including logging fixes in next iteration.

**Comment by sandeepmjay** [comment]
Thanks @johncrossman for the review. Will be including logging related fixes in the next refactor. Please merge the changes at your convenience.

**Comment by johncrossman** [comment]
Thank you for the replies, @sandeepmjay  Merge it!

**Pull request merged**